expyfun/stimuli/_tracker.py

```diff
@@ -0,0 +1,837 @@
+    parameter must always be the same as that of ``change_criteria``, and the
+    first element of change_criteria must always be 0.
+    For the example given above:
+        ``step_size_up=[1., 0.2]``, ``step_size_down=[1., 0.2]``,
+        ``change_criteria=[0, 2]``, ``change_rule='reversals'``
+    would change the step sizes from 1 to 0.2 after two reversals.
+
+    If static step sizes are used, both ``step_size_up``
+    and ``step_size_up`` must be scalars and ``change_criteria`` must be None.
+    """
+    def __init__(self, callback, up, down, step_size_up, step_size_down,
+                 stop_criterion, stop_rule, start_value, change_criteria=None,
+                 change_rule='reversals', x_min=None, x_max=None):
+        self._callback = _check_callback(callback)
+        self._up = up
+        self._down = down
+        self._stop_criterion = stop_criterion
+        self._stop_rule = stop_rule
+        self._start_value = start_value
+        self._x_min = x_min
+        self._x_max = x_max
+
+        if change_criteria is None:
+            change_criteria = [0]
+        elif change_criteria[0] != 0:
+            raise(ValueError('First element of change_criteria must be 0.'))
+        self._change_criteria = np.asarray(change_criteria)
+        if change_rule not in ['trials', 'reversals']:
+            raise(ValueError, "change_rule must be either 'trials' or "
+                  "'reversals'")
+        self._change_rule = change_rule
+
+        step_size_up = list(np.atleast_1d(step_size_up))
+        if len(step_size_up) != len(change_criteria):
+            raise(ValueError('If step_size_up is not scalar it must be the '
+                             'same length as change_criteria.'))
+        self._step_size_up = np.asarray(step_size_up, dtype=float)
+
+        step_size_down = list(np.atleast_1d(step_size_down))
+        if len(step_size_down) != len(change_criteria):
+            raise(ValueError('If step_size_down is not scalar it must be the '
+                  'same length as change_criteria.'))
+        self._step_size_down = np.asarray(step_size_down, dtype=float)
+
+        self._x = np.asarray([start_value], dtype=float)
+        self._x_current = start_value
+        self._responses = np.asarray([], dtype=bool)
+        self._reversals = np.asarray([], dtype=int)
+        self._n_up = 0
+        self._n_down = 0
+
+        self._direction = 0
+        self._n_trials = 0
+        self._n_reversals = 0
+        self._stopped = False
+
+        # Now write the initialization data out
+        self._tracker_id = id(self)
+        self._callback('tracker_identify', json.dumps(dict(
+            tracker_id=self._tracker_id,
+            tracker_type='TrackerUD')))
+
+        self._callback('tracker_%i_init' % self._tracker_id, json.dumps(dict(
+            callback=None,
+            up=up,
+            down=down,
+            step_size_up=step_size_up,
```

Comment:
Ah, sure that's better. I was thinking I would have to modify the callback itself, but of course that's not true.